lib/maths/CBoostedTreeLeafNodeStatistics.cc

```diff
@@ -325,68 +275,134 @@ CBoostedTreeLeafNodeStatistics::computeBestSplitStatistics(const TRegularization
 
     SSplitStatistics result;
 
-    for (auto i : featureBag) {
-        std::size_t c{m_MissingDerivatives[i].s_Count};
-        double g{m_MissingDerivatives[i].s_Gradient};
-        double h{m_MissingDerivatives[i].s_Curvature};
-        for (const auto& derivatives : m_Derivatives[i]) {
-            c += derivatives.s_Count;
-            g += derivatives.s_Gradient;
-            h += derivatives.s_Curvature;
+    // We seek to find the value at the minimum of the quadratic expansion of the
+    // regularized loss function. For a given leaf this expansion is
+    //
+    //   L(w) = 1/2 w^t H(\lambda) w + g^t w
+    //
+    // where H(\lambda) = \sum_i H_i + \lambda I, g = \sum_i g_i and w is the leaf's
+    // weight. Here, g_i and H_i denote an example's loss gradient and Hessian and i
+    // ranges over the examples in the leaf. Completing the square and noting that
+    // the minimum is given when the quadratic form is zero we have
+    //
+    //   L(w^*) = -1/2 g^t H(\lambda) g
```

Comment:
```suggestion
    //   L(w^*) = -1/2 g^t H(\lambda)^{-1} g
```
I guess, you need to invert the Hessian.